tiledb/sm/filter/webp_filter.cc

```diff
@@ -0,0 +1,324 @@
+  throw_if_not_ok(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  int extent_y = extents_.first, extent_x = extents_.second,
+      pixel_depth = format_ < WebpInputFormat::WEBP_RGBA ? 3 : 4;
+  // X should be divisible by colorspace pixel_depth or RGB values will skew.
+  if (extent_x % pixel_depth != 0) {
+    throw StatusException(Status_FilterError(
+        pixel_depth == 3 ?
+            "Colorspace with no alpha must use extents divisible by 3" :
+            "Colorspace with alpha must use extents divisible by 4"));
+  }
+
+  for (const auto& i : input_parts) {
+    auto data = static_cast<const uint8_t*>(i.data());
+    // Number of bytes encoded; Encoded result data buffer.
+    size_t enc_size = 0;
+    uint8_t* result = nullptr;
+    // Cleanup allocated data when we leave scope.
+    ScopedExecutor cleanup([&]() {
+      if (enc_size > 0 && result != nullptr)
+        WebPFree(result);
+    });
+    // We divide extent_x by colorspace value count to get pixel-width of image,
+    // extent_x currently represents row stride.
+    switch (format_) {
+      case WebpInputFormat::WEBP_RGB:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessRGB(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeRGB(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_RGBA:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessRGBA(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeRGBA(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_BGR:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessBGR(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeBGR(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_BGRA:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessBGRA(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeBGRA(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_NONE:
+        throw StatusException(Status_FilterError(
+            "Filter option TILEDB_FILTER_WEBP_FORMAT must be set"));
```

Comment:
Do we need a `default` here? I guess if nothing matches then we will hit the next `if` error block, but that makes an assumption about the code structure which might not hold in the future. So, definsively, a `default` case may make sense.